### ISSUE TITLE
Question of the paper——The forward pass may forget one rescaling？

### ISSUE THREAD

**Issue by Wu-Fisher** [Dao-AILab/flash-attention#411]
On page 4 of flashattention2, is the 4th equation of O(1) has the element $$e^{S(1)-m(1)}$$
but the last equation just becomes the $$e^{S(1)-m}$$
This happened on page 5, too.
From the former paper flashattention1, it has a rescaling of $$e^{m(1)-m} $$ in the equation of O(2) and O(1). 
Is there a clerical error？ （Overall, the two papers really attract me!!!）

**Comment by jundaf2** [comment]
Hi @Wu-Fisher, I think it is not an error.

m_new = (m_old, m_temporary), the m(1) is m_old and m is m_new here.

Correct me if I am wrong. Thanks

**Comment by Wu-Fisher** [comment]
@jundaf2 yep >😂 m_new = (m_old, m_temporary), the m(1) is m_old and m is m_new here.
But the O(2) 's equation just uses m to replace m(1) (flashattention2 page4 around O(1) ). 
From the flashattention1, O(2) = diag(l1/l2)^(-1)e(m-m1)O(1)....

**Comment by Wu-Fisher** [comment]
@jundaf2 I am not sure what the sentence "the first term in the last line of page 2" means. (I didn't find it in both two papers)
I can't find which equation on page 4 shows that O1 has been scaled. 
Of course, it is right if O1 has been scaled.

**Comment by jundaf2** [comment]
> 

Sorry, I mean page 4. Maybe it's better to let Dao to confirm.

**Comment by Wu-Fisher** [comment]
@jundaf2 Thx🙂,  I think it was just a clerical error or a lack of rigor. Equations on Page 6 are right as you can see diag(m(1)-m(2))O1.

**Comment by eggiter** [comment]
- My thought of the right equation is:
$$diag({l^{(2)}}/{l^{(1)}})^{-1}·e^{m(1)-m(2)}·O(1)$$

**Comment by hjyai94** [comment]
> * My thought of the right equation should be:
>   diag(l(2)/l(1))−1·em(1)−m(2)·O(1)
I also think the right format of equation shoud be :

$$ O^{(2)} = \text{diag}\left(\frac{l^{(1)}}{l^{(2)}} \right)  e^{m^{(1)}-m^{(2)}}O^{(1)} + \tilde{P}^{(2)} V^{(2)} $$

as mentioned in https://github.com/Dao-AILab/flash-attention/issues/897#issue-2202994515
